### ISSUE TITLE
Accuracy drop for compiled MobileNetV2 model

### ISSUE THREAD

**Issue by fcakir** [google-coral/edgetpu#77]
Hi, I'm seeing performance accuracy drops if I try to compile (via edgetpu_compiler) a mobilenet V2 model for image classification.

 When I use the pre-compiled MobileNetV2 model provided at https://coral.ai/models/, the imagenet validation accuracy I get on the TPU is 71.3%. But when I do the compilation myself with the [quantized MobileNetV2 TFlite model](https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/mobilenet_v2_1.0_224_quant.tgz) the performance drops significantly (below 10%). I'm using the same preprocessing steps. The labels are the same as well. I've also tested the quantized TFLite model on CPU and its accuracy is fine.

**Comment by Namburger** [comment]
@fcakir what compiler version are you using and how are you testing this performance?
A model after compiling shouldn't cause any changes in the results at all. Take this very basic POC script for instance:

```
import numpy as np
import sys
from tflite_runtime.interpreter import Interpreter
from tflite_runtime.interpreter import load_delegate

def main():
    model_path = 'mobilenet_v2_1.0_224_quant.tflite'
    print('cpu model')
    interpreter = Interpreter(model_path)
    interpreter.allocate_tensors()
    input_details = interpreter.get_input_details()
    image = np.zeros(input_details[0]['shape'], input_details[0]['dtype'])
    interpreter.set_tensor(input_details[0]['index'], image)
    interpreter.invoke()
    output_details = interpreter.get_output_details()
    outputs1 = interpreter.get_tensor(output_details[0]['index'])
    print(outputs1)

    model_path = 'mobilenet_v2_1.0_224_quant_edgetpu.tflite'
    print('edgetpu model')
    delegates = [load_delegate('libedgetpu.so.1.0')]
    interpreter2 = Interpreter(model_path, experimental_delegates=delegates)
    interpreter2.allocate_tensors()
    input_details2 = interpreter2.get_input_details()
    interpreter2.set_tensor(input_details2[0]['index'], image) # using same inputs
    interpreter2.invoke()
    output_details = interpreter2.get_output_details()
    outputs2 = interpreter2.get_tensor(output_details[0]['index'])
    print(outputs2)

    assert outputs1.all() == outputs2.all()

    print('Success.')

if __name__== '__main__':
    main()
```

I'm running both the CPU and the TPU model on the same inputs and getting the same result! 
I'd definitely check to see if you are loading the correct model on your test

**Comment by tigert1998** [comment]
I met the same problem. Currently, I am using TensorFlow 2.1 to do the integer quantization and apply the latest edgetpu compiler (2.0.291256449) to compile it. But I got an extremely low accuracy (< 10%).

I know that the official website said tf 1.15 is a must to quantize. But the compiler said all the ops except the very first and last quant and dequant op are still running on tpu. So I think it might be OK?

Anyway, I am curious why this would happen. Is anybody can explain this to me?
Also, since tf 1.15 make inputs and outputs all in int8 type, **how can I correctly preprocess the image**?

**Comment by Namburger** [comment]
@tigert1998 thanks for the EDIT lol
I got really nervous for a second!

**Comment by tigert1998** [comment]
Hi @Namburger:
I am using TF 2.1 just the same as the last post. I found some edgetpu model (post int8 quantization and edgetpu compiler) behaviors are not consistent with those before compiling (just after post int8 quantization), like Mnasnet and Inception V4.

I want to ask that whether tf 2.1 can be used to do the int8 quantization for edgetpu or not?

**Comment by Namburger** [comment]
@tigert1998 You can use tf2.1 for post training quantization but with tensorflow 2.0, the option for setting input/output tensors are deprecated. That means you'll need to pre process your data before running inference. This is why we still suggest users to use tf1.15 for ptq but our current release of the compiler will let you compiler models with float input/output as long as intermediate tensors are quantized. Could you also attach an example model here that gives incorrect behavior for us to evaluate?

**Comment by tigert1998** [comment]
Thank @Namburger so much for your rapid response!

Mnasnet post int quant and edgetpu version:
https://1drv.ms/u/s!Ame-g9xGXIZyhmPtAnWjze9nFo5t?e=A5vB4S

The model and its weights come from google's official repo:
https://github.com/tensorflow/tpu/blob/master/models/official/mnasnet/mnasnet_main.py

This model uses the normalization: input = (original_image - mean) / std, where mean = [123.67500305175781, 116.27999877929688, 103.52999877929688] and std = [58.39500045776367, 57.119998931884766, 57.375].

With tf lite 2.1, I achieve a 72+% ImageNet validation accuracy. However, with edgetpu, the number goes down to 8+%.

**Comment by Namburger** [comment]
Humnn, so regardless of the normalization, which you should be taking care of while processing the image before run inference. Otherwise, with random inputs, your tflite model should behave exactly like your edgetpu model, it's quite concerning that it's not. Previously we see these issues only with cocompiled models, taking a look at yours now.

**Comment by Namburger** [comment]
@tigert1998 So actually, I'm quite confused on this situation here o_0
Based on your query:
```
2. And if I use tf1.15, how can I do the correct preprocess for the model that only accepts a int8 image (since many models use different normalization strategies, and I cannot apply the preprocess if input type is just int8)?
```
Your model was already quantized? So you performed post training quantization on a model that's already quantized and it produces a model that takes float inputs?

**Comment by tigert1998** [comment]
@Namburger 
Now I get your confusion. The word "preprocess" is overloaded between you and me. My "preprocess" is the "preprocess" for input data (image), and it is all about center crop and normalization. So don't understand "preprocess" as a model compiling, OK?  😄 

My two uploaded model:
1. the one whose name ends with "_int_quant": post int quantized with tflite 2.1
2. the one whose name ends with "_edgetpu": the last model additionally compiled with edgetpu compiler

**Comment by Namburger** [comment]
@tigert1998 so I actually understand prepocess in your context 🙂
What I meant is that if you use post training quantization with tf2*, then your I outs tensors will be of type float and input will need to be prepocessed before you feed it to the interpreter, [like this example](https://github.com/google-coral/edgetpu/blob/master/src/cpp/basic/float_input_model_test.cc#L19) (btw this new feature came out yesterday, not extended for the python api yet tho).

I also can distaguised between both of your models. What I'm wondering is if the original tf model was already quantized before you applied post training quantization?

**Comment by tigert1998** [comment]
1. I did do the correct preprocess for the float inputs and outputs.
2. I don't even know how to quantize a model twice. I did only two steps: I created a frozen pb whose weights are directly imported from the official model repo. And then I made post training quantization to it. That's all.

**Comment by Namburger** [comment]
@tigert1998 
Okay, just wanted to clarify that, since I couldn't quite understand your original query and I really thought you said your model only take in uint8 inputs before applying post training quantization :).

> Did float preprocessing for edgetpu inference just officially come out yesterday? It sounds much later than I thought it would be. So the aforementioned python code I wrote for inference is invalid since python api is not ready?
Sorry for confusion, this came out yesterday for our [edgetpu api](https://github.com/google-coral/edgetpu/tree/master/edgetpu). If you are using the `tflite_runtime` API, which came from the tensorflow team, you always had that options. To clarify, your code looks good :) I'll past on your models to my team today for some further analysis.

**Comment by Namburger** [comment]
@tigert1998 What preprocess library are you using for `preprocess.execute(image_path)`?
> With tf lite 2.1, I achieve a 72+% ImageNet validation accuracy. However, with edgetpu, the number goes down to 8+%.

Could you by any chance send me an example inference script with a label file to validate this finding?
